From 3c733c78d6f8e50296d6e248968b6516c92a53ca Mon Sep 17 00:00:00 2001
From: Natalia <124304+nessita@users.noreply.github.com>
Date: Mon, 19 Aug 2024 14:47:38 -0300
Subject: [PATCH] [5.1.x] Fixed CVE-2024-45231 -- Avoided server error on
 password reset when email sending fails.

On successful submission of a password reset request, an email is sent
to the accounts known to the system. If sending this email fails (due to
email backend misconfiguration, service provider outage, network issues,
etc.), an attacker might exploit this by detecting which password reset
requests succeed and which ones generate a 500 error response.

Thanks to Thibaut Spriet for the report, and to Mariusz Felisiak, Adam
Johnson, and Sarah Boyce for the reviews.
---
 django/contrib/auth/forms.py   |  9 ++++++++-
 docs/ref/logging.txt           | 12 ++++++++++++
 docs/releases/4.2.16.txt       | 11 +++++++++++
 docs/releases/5.0.9.txt        | 11 +++++++++++
 docs/releases/5.1.1.txt        | 11 +++++++++++
 docs/topics/auth/default.txt   |  4 +++-
 tests/auth_tests/test_forms.py | 21 +++++++++++++++++++++
 tests/mail/custombackend.py    |  6 ++++++
 8 files changed, 83 insertions(+), 2 deletions(-)

diff --git a/django/contrib/auth/forms.py b/django/contrib/auth/forms.py
index a11668944a82..edf672a6e516 100644
--- a/django/contrib/auth/forms.py
+++ b/django/contrib/auth/forms.py
@@ -418,7 +420,12 @@ def send_mail(
             html_email = loader.render_to_string(html_email_template_name, context)
             email_message.attach_alternative(html_email, "text/html")
 
-        email_message.send()
+        try:
+            email_message.send()
+        except Exception:
+            logger.exception(
+                "Failed to send password reset email to %s", context["user"].pk
+            )
 
     def get_users(self, email):
         """Given an email, return matching user(s) who should receive a reset.
